### ISSUE TITLE
Blazor component route inheritance to view (.cshtml)

### ISSUE THREAD

**Issue by rafek1241** [aspnet/Blazor#908]
Since component is not a partial class the issue is that when i specify route (e.g. `/interop`) in base class and i won't use `@page` in the view but use `@inherits` - i've got issue in console in the browser: 
```
MonoPlatform.ts:70
 Uncaught (in promise) Error: System.InvalidOperationException: Failed to compare two elements in the array. ---> System.InvalidOperationException: The following routes are ambiguous:
'interop' in 'blazorfirststandaloneapplication.Base.InteropBase'
'interop' in 'blazor_first_standalone_application.Pages.Interop'
```

If it's inheritance - in my opinion it should derive Route attribute.

**Comment by Andrzej-W** [comment]
I don't agree with you. I think there are only two reasons why you want to use base class and inheritance:
1. You want to have two separate files for a single component (`cshtml` and `cs`). In that case you can specify `@page` in the view. Why do you want to do this in `cs` file?
2. You want to have a base class for a few derived components. In that case you have to specify `@page` in derived component because you cannot have 2 different components with the same route. Once again you can have `@page` in the views.

Am I missing something?

**Comment by rafek1241** [comment]
No - you don't. Thank you for your answer. From what you have written I can assume that when I will add an [Route()] attribute to the component base class then this class will be excluded from inheritance.
Is that correct?

**Comment by Andrzej-W** [comment]
I'm sorry, but I don't understand what are you going to achieve.

**Comment by rafek1241** [comment]
I was reading https://learn-blazor.com/pages/dynamic-content/ and from that page I have found that we can create components with routes but only when it's code-only. I don't want to convince you to my opinion- I just want to understand why it is like that and not for example - some special class which derive from BlazorComponent and enables Route attribute because thanks to you I have found that it is working for now only in this way. Earlier I wanted to remove `@page` directives just to make my views more html-like. I wanted to achieve only one directive in the first line `@inherits`. Now I know that i can't achieve that.

**Comment by Andrzej-W** [comment]
Sorry, I have absolutely no experience with code only components.

**Comment by dlr1** [comment]
I tried the following and expected it to work like @rafek1241 says., with daily builds, it doesn't work, but if I move the route to .cshtml it works fine. I don't see any errors in the console
```
[Route("/test")]
    public class Testbase : BlazorComponent
    {
       
    }
```
the following in .cshtml
```
@inherits Testbase
this is a test

**Comment by dlr1** [comment]
On the other hand, if the page has multiple levels of inheritance, it doesn't make sense to have the route attribute on the base classes, so to avoid I guess it makes sense to have it in .cshtml only.

**Comment by rafek1241** [comment]
So you agree with me or not?
I made new repository just to show you the problem. When you load this project you will see error inside developer tools console in Browser.
[routing-problem-blazor](https://github.com/rafek1241/routing-problem-blazor)

**Comment by dlr1** [comment]
@rafek1241 As I said in my first post, I have expected it to work. the gist of what you have is the same as the code I posted above. 

But consider this scenario. lets say, there is code only class which has a route specified on it, like the url you gave above and what if 2 pages like in your example(Counter1 and Counter2) inherit from that code only component and have Route attribute applied to them as well. should the router be routing to base class  or the inherited class. how does routing work in this case?

**Comment by SteveSandersonMS** [comment]
Currently the `RouteTable` does not look for inherited `RouteAttribute` on components: https://github.com/aspnet/Blazor/blob/dev/src/Microsoft.AspNetCore.Blazor/Routing/RouteTable.cs#L26

This explains what @rafek1241 is observing. Flagging as a bug.

Current workaround is to do what you've noted above, which is to put the route information on the derived class (e.g., in the `.cshtml` file).

**Comment by dlr1** [comment]
@SteveSandersonMS how would the routing work in the case I mentioned above?

**Comment by Andrzej-W** [comment]
@dlr1 I suppose that inherited `RouteAttribute` will be checked only if there is no attribute or `@page` directive in derived class. On the other hand if there are two derived components ***without*** `RouteAttribute` or `@page` directive we should probably get an exception at application startup. Currently when two components have the same url in `@page` directive we get a nice exception:
```
Error: System.InvalidOperationException: Failed to compare two elements in the array. ---> System.InvalidOperationException: The following routes are ambiguous:
'counter' in 'Spa.Client.Pages.Counter2'
'counter' in 'Spa.Client.Pages.Counter'
```

**Comment by dlr1** [comment]
We don't need to have Route attributes, if the components are not going to be routed, but used in other components

**Comment by rafek1241** [comment]
-----------
The workaround and solution for that (first):
1. Remove allow for inheritance in the RouteAttribute.
2. Create another class which derive from BlazorComponent and allow use RouteAttribute only for this class. (That is how we can create code-only components).
3.Modify compiler that blazor component layout `.cshtml` can derive only from blazorComponent (not from the code-only class)
Result: 
-Pros: We don't have [Route()] in blazor component so user can't setup route in the blazor component layout and blazor component class in the same time.
-Cons: We don't have choose where we store our routing.
----------
Second solution:
Change interpretation of this attribute in compiler. Attribute shouldn't be inherited or if it is - `@page` directive should be unnecessery when you derive from component class which are using RouteAttribute and we should be informed in IDE that base component class is trying to override our `@page` directive (as ERROR).
